daphne/src/messages.rs

```diff
@@ -809,6 +817,218 @@ impl Decode for HpkeCiphertext {
+    None,
+    NotImplemented(u8),
+}
+
+impl From<DpMechanism> for u8 {
+    fn from(dp_mech: DpMechanism) -> Self {
+        match dp_mech {
+            DpMechanism::Reserved => DP_MECHANISM_RESERVED,
+            DpMechanism::None => DP_MECHANISM_NONE,
+            DpMechanism::NotImplemented(x) => x,
+        }
+    }
+}
+
+impl Encode for DpMechanism {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        u8::from(*self).encode(bytes);
+    }
+}
+
+impl Decode for DpMechanism {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        match u8::decode(bytes)? {
+            x if x == DP_MECHANISM_RESERVED => Ok(Self::Reserved),
+            x if x == DP_MECHANISM_NONE => Ok(Self::None),
+            x => Ok(Self::NotImplemented(x)),
+        }
+    }
+}
+
+// The Differential Privacy configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct DpConfig {
+    pub mechanism: DpMechanism,
+}
+
+impl AsRef<DpConfig> for DpConfig {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl Encode for DpConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.mechanism.encode(bytes);
+    }
+}
+
+impl Decode for DpConfig {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            mechanism: DpMechanism::decode(bytes)?,
+        })
+    }
+}
+
+// The VDAF configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct VdafConfig {
+    pub dp_config: DpConfig,
+    pub vdaf_type: VdafType,
+}
+
+impl AsRef<VdafConfig> for VdafConfig {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl Encode for VdafConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.dp_config.encode(bytes);
+        self.vdaf_type.encode(bytes);
+    }
+}
+
+impl Decode for VdafConfig {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            dp_config: DpConfig::decode(bytes)?,
+            vdaf_type: VdafType::decode(bytes)?,
+        })
+    }
+}
+
+// URL encode / decode helper struct
+struct UrlBytes {
+    bytes: Vec<u8>,
+}
+
+impl UrlBytes {
+    fn to_url(&self) -> Result<Url, CodecError> {
+        let s = from_utf8(&self.bytes).or_else(|e| Err(CodecError::Other(e.into())))?;
+        Url::parse(&s).or_else(|e| Err(CodecError::Other(e.into())))
+    }
+}
+
+impl Encode for UrlBytes {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        encode_u8_items(bytes, &(), &self.bytes)
+    }
+}
+
+impl Decode for UrlBytes {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            bytes: decode_u8_items(&(), bytes)?,
+        })
+    }
+}
+
+// The Task configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct TaskConfig {
+    pub task_info: Vec<u8>,
+    pub aggregator_endpoints: Vec<Url>,
```

Comment:
Hmm, so this would mean handling a malformed URL as a CodeError, right? I don't think this is quite what we want. In the actual protocol, the helper would end up aborting with "unrecognizedMessage", which isn't a helpful feedback for the leader.

I would treat this as opaque bytes here and leave it to the application to parse the bytes as a URL.
```suggestion
    pub aggregator_endpoints: Vec<Vec<u8>>,
```

Reply:
ok, much easier for me!

Reply:
Url simplified to Vec<u8>